content/blog/next-level-iac-briding-the-declarative-gap/index.md

```diff
@@ -0,0 +1,166 @@
+---
+
+Pulumi stands out in the world of infrastructure-as-code (IaC) for its flexibility and ease of use. This is due to its unique approach of using general purpose programming languages to describe system configuration. It accomplishes this through a small but brilliant implementation detail that bridges the worlds of declarative and standard programming using asynchronous semantics.
+
+<!--more-->
+
+In Pulumi, you describe your infrastructure in code – real code, not a [DSL][dsl-wiki] or [YAML][yaml-wiki] – using your preferred general purpose programming language You don't have to become a specialist in a niche proprietary declarative language, like Terraform's HCL. Rather, you can reuse your existing programming skills, writing in standard imperative, object-oriented, and even functional language styles, while still gaining all of the benefits of the declarative style that other tools emphasize.
+
+Pulumi provides a unique mix of a [declarative model][pulumi-declarative-imperative-docs] embedded and implemented inside of a standard programming language, allowing all the flexibility of custom imperative code, while still enabling Pulumi's [deployment engine][pulumi-engine-docs] to infer opportunities for parallel asynchronous execution and to converge a partially-realized system.
+
+The secret sauce that lets Pulumi achieve this is the [`apply`][apply-docs] function, and the way it handles inputs and outputs via asynchronous programming constructs, similar to [futures or promises][futures-wiki]. We'll go deeper on that in this article, but first let's talk a little bit about the flow of operations in Pulumi.
+
+## Pulumi's architecture and flow of operations
+
+Pulumi is a system that is made up of a number of components. At a high level, those are:
+
+- [**Pulumi Programs**][pulumi-program-docs]: The code you write to describe the infrastructure you need.
+- [**Deployment Engine**][pulumi-engine-docs]: A process that executes requests to create/modify infrastructure.
+- [**Resource Providers**][pulumi-providers-docs]: Plugins that interface with cloud services like AWS, Azure, and others.
+
+These components interact together to manage your infrastructure across multiple cloud providers using a consistent programming and execution model.
+
+{{< figure src="pulumi-architecture-simplified.png" caption="Figure: Pulumi's Program, Deployment Engine, and Resource Provider architecture">}}
+
+Let's get into an example of how information flows back forth between these components. Suppose you want to write a Pulumi program to create an AWS S3 Bucket to [host a static website][pulumi-static-website-example]. When that bucket is created, it will be given a unique indentifier that you need to use in other parts of your infrastructure. Every time you create the bucket the identifier will change, and it's generated by the AWS API, so we can't know it ahead of time. We need to handle this as a variable, not a fixed configuration value.
```

Comment:
```suggestion
Let's get into an example of how information flows back and forth between these components. Suppose you want to write a Pulumi program to create an AWS S3 Bucket to [host a static website][pulumi-static-website-example]. When that bucket is created, it will be given a unique indentifier that you need to use in other parts of your infrastructure. Every time you create the bucket the identifier will change, and it's generated by the AWS API, so we can't know it ahead of time. We need to handle this as a variable, not a fixed configuration value.
```